### PR TITLE
fix: evaluate expression from root

### LINKED ISSUE #66: Expression tree should not be evaluated if logical operator already satisfies
Example should only evaluate left hand side if lateUpdatedDatalake is '':
@or(equals(item().lastUpdatedDatalake,''),greaterOrEquals(item().lastUpdatedSourceSystem,addSeconds(item().lastUpdatedDatalake,-1)))

### PR DESCRIPTION
**Pull request by LeonardHd**
Refactors the evaluation of expressions to allow evaluation from root to leaves.

- Splits concerns into different transformers,
- Instead of transformers rule logic is contained in decorator patterns,
- Expression tree is evaluated from top to bottom (recursively).

Fixes #66

### CODE REVIEW

File: src/data_factory_testing_framework/functions/evaluator/expression_evaluator.py [hunk 1]
```diff
@@ -39,7 +46,8 @@ def __init__(self) -> None:
         expression_grammar = f"""
             // TODO: add support for array index
             ?expression_start: expression_evaluation
-            expression_evaluation: expression_call [expression_object_accessor]*
+            # TODO: probably object accessor does not apply to all below in expression_evaluation
+            expression_evaluation: (expression_logical_bool | expression_branch | expression_call) ((("." EXPRESSION_PARAMETER_NAME) | EXPRESSION_ARRAY_INDEX)+)?
```
Comment (arjendev) [inline src/data_factory_testing_framework/functions/evaluator/expression_evaluator.py:50]: What's the reason of merging expression_calls with expression_accessors in a single rule?

File: src/data_factory_testing_framework/functions/evaluator/rules/evaluation_expression_rule.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+                f"Invalid number of children. Minimum required: 1, Actual: {len(self.children)}"
+            )
+
+        if not isinstance(self.children[0], ExpressionRuleEvaluator):
+            raise ExpressionEvaluationInvalidChildTypeError(
+                child_index=0, expected_types=ExpressionRuleEvaluator, actual_type=type(self.children[0])
+            )
+
+    def evaluate(self) -> EvaluatedExpression:
+        expression = self.children[0]
```
Comment (arjendev) [inline src/data_factory_testing_framework/functions/evaluator/rules/evaluation_expression_rule.py:26]: Maybe we should move the unpacking of children to the constructor and also do all the type checks there, so that the evaluate method is purely about evaluating and the constructor about constructing and validating?

Comment (LeonardHd) [reply src/data_factory_testing_framework/functions/evaluator/rules/evaluation_expression_rule.py:26]: Good idea actually!

File: src/data_factory_testing_framework/functions/evaluator/rules/expression_rule.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+from dataclasses import dataclass
+from typing import Any, Union
+
+from lark import Tree
+
+from data_factory_testing_framework.functions.evaluator.exceptions import ExpressionEvaluationInvalidChildTypeError
+
+
+@dataclass
+class EvaluatedExpression:
```
Comment (arjendev) [inline src/data_factory_testing_framework/functions/evaluator/rules/expression_rule.py:11]: The name `EvaluatedExpression` feels a bit counter-intuitive in a number of places. It can represent an evaluated expression, but it can also represent a plain literal. Maybe we can brainstorm about this one together.

File: src/data_factory_testing_framework/functions/evaluator/expression_evaluator.py (hunk 1 above)
Comment (LeonardHd) [reply src/data_factory_testing_framework/functions/evaluator/expression_evaluator.py:50]: Good remark, and I hope my memory doesn't fail me:
Now that we evaluate the three from top to bottom the `expression_object_accessor` needs to have the actual values it accesses as children. The tree composition therefore changes. There are likely multiple ways to achieve it, but I made it part of the `expression_evaluation` mostly to keep the evaluated tree as simple as possible to keep it maintainable and DRY. 

Previously, the `expression_evaluation` rule was dead simple (just a passthrough afaik) but also used to to the expression_object_accesor processing:
```python
def expression_evaluation(self, values: list[Token, str, int, float, bool, list]) -> [str, int, float, bool]:
        eval_value = values[0]

        remaining_fields = values[1:]
        return self._evaluate_expression_object_accessors(eval_value, remaining_fields)
```

Bottom line: the changes just simplify the tree, but things stay (more or less) the same.
For me unclear, but also present in the code base, is if all other rules acutally support arrays or objects, this left a TODO:
`# TODO: probably object accessor does not apply to all below in expression_evaluation`. If we, however, assume valid ADF expressions that this should not matter (for now)

File: src/data_factory_testing_framework/functions/evaluator/rules/expression_rule.py (hunk 3 above)
Comment (LeonardHd) [reply src/data_factory_testing_framework/functions/evaluator/rules/expression_rule.py:11]: Yes - wasn't really pleased with the naming either. Indeed, it represents literal values and evaluated values. In the end it's actual values that are derived from evaluating the parsed lark tree, thus my naming. Happy to hear suggestions.

File: src/data_factory_testing_framework/functions/evaluator/rules/logical_bool_expression_rule.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    def _check_child_type(self, child: Union[EvaluatedExpression, ExpressionRuleEvaluator], child_index: int) -> None:
+        if not isinstance(child, (ExpressionRuleEvaluator, EvaluatedExpression)):
+            raise ExpressionEvaluationInvalidChildTypeError(
+                child_index=child_index,
+                expected_types=(ExpressionRuleEvaluator, EvaluatedExpression),
+                actual_type=type(child),
+            )
+
+    def _evaluate_expression(self, expression: Union[ExpressionRuleEvaluator, EvaluatedExpression]) -> bool:
+        if isinstance(expression, ExpressionRuleEvaluator):
```
Comment (arjendev) [inline src/data_factory_testing_framework/functions/evaluator/rules/logical_bool_expression_rule.py:57]: I noticed that the pattern of instance-checking RuleEvaluator and EvaluatedExpression and then either calling evaluate() or .value respectively is happening a lot. Perhaps we can abstract this away in a single place, hopefully in the class itself?

File: tests/unit/functions/test_expression_evaluator.py [hunk 5]
```diff
@@ -7,493 +7,16 @@
...
-        p("value", Tree(Token("RULE", "literal_evaluation"), [Token("LITERAL_LETTER", "value")]), id="string_literal"),
-        p(
-            " value ",
-            Tree(Token("RULE", "literal_evaluation"), [Token("LITERAL_LETTER", "value")]),
-            id="string_with_ws_literal",
-            marks=pytest.mark.skip(""),
-        ),
-        p("11", Tree(Token("RULE", "literal_evaluation"), [Token("LITERAL_INT", "11")]), id="integer_literal"),
-        p(
-            "@pipeline().parameters.parameter",
```
Comment (arjendev) [inline tests/unit/functions/test_expression_evaluator.py:33]: Any reason on removing all these tests that validate whether an expression is resulting in the correct tree?

**Review by arjendev — CHANGES_REQUESTED** [review]
Looks neat @LeonardHd and solves an important issue! I have added a few remarks.

File: src/data_factory_testing_framework/functions/evaluator/rules/logical_bool_expression_rule.py (hunk 4 above)
Comment (LeonardHd) [reply src/data_factory_testing_framework/functions/evaluator/rules/logical_bool_expression_rule.py:57]: We are getting there :100: ! Named it for now `ensure_evaluated_expression` and let's find a better name once we also adjust `EvaluatedExpression`.

File: src/data_factory_testing_framework/functions/evaluator/rules/activity_reference_expression_rule.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+        if not isinstance(self.children[0], EvaluatedExpression):
+            raise ExpressionEvaluationInvalidChildTypeError(
+                child_index=0, expected_types=EvaluatedExpression, actual_type=type(self.children[0])
+            )
+
+        self.activity_name = self.children[0].value
+
+    def evaluate(self) -> EvaluatedExpression:
+        activity = self.state.get_activity_result_by_name(self.activity_name)
```
Comment (arjendev) [inline src/data_factory_testing_framework/functions/evaluator/rules/activity_reference_expression_rule.py:30]: activity_result

File: src/data_factory_testing_framework/functions/evaluator/rules/expression_rule.py (hunk 3 above)
Comment (arjendev) [reply src/data_factory_testing_framework/functions/evaluator/rules/expression_rule.py:11]: Suggestion of brainstorm: EvaluationResult

File: src/data_factory_testing_framework/functions/evaluator/rules/expression_rule.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    def evaluate(self) -> EvaluatedExpression:
+        """Evaluates the expression rule."""
+        pass
+
+    @property
+    def children(self) -> list[Union["ExpressionRuleEvaluator", EvaluatedExpression]]:
+        """Returns the children of the expression rule."""
+        return self._tree.children
+
+    def ensure_evaluated_expression(
```
Comment (arjendev) [inline src/data_factory_testing_framework/functions/evaluator/rules/expression_rule.py:46]:
```suggestion
    def evaluate_child(
```

File: src/data_factory_testing_framework/functions/evaluator/rules/linked_service_reference_expression_rule.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+        if not isinstance(self.children[0], EvaluatedExpression):
+            raise ExpressionEvaluationInvalidChildTypeError(
+                child_index=0, expected_types=EvaluatedExpression, actual_type=type(self.children[0])
+            )
+
+        self.parameter_name = self.children[0].value
+
+    def evaluate(self) -> EvaluatedExpression:
+        activity = self.state.get_parameter_by_type_and_name(RunParameterType.LinkedService, self.parameter_name)
```
Comment (arjendev) [inline src/data_factory_testing_framework/functions/evaluator/rules/linked_service_reference_expression_rule.py:31]: parameter

**Review by arjendev — CHANGES_REQUESTED** [review]
Few minor comments outstanding

**Comment by LeonardHd** [comment]
> Few minor comments outstanding

Addressed the outstanding issues :+1:

**Review by arjendev — APPROVED** [review]
Awesome changes, approved!
